File: esm/engines/02_memory/scar.js
/**
 * Sovereign Critical Action Record (SCAR) database and Repeat Failure Guard.
 * Prevents AI agents from repeating identical errors in loops.
 * Upgraded to ESModules with seamless in-memory fallback.
 */

import fs from 'fs';
import path from 'path';

let DatabaseModule = null;
let db = null;
let useFallback = false;
let fallbackIncidents = [];
const fallbackFile = path.resolve(process.cwd(), 'data', 'incidents.json');

try {
  const betterSqlite = await import('better-sqlite3');
  DatabaseModule = betterSqlite.default;
  const dbPath = path.resolve(process.cwd(), 'data', 'scars.db');
  fs.mkdirSync(path.dirname(dbPath), { recursive: true });
  db = new DatabaseModule(dbPath);
  db.exec(`
    CREATE TABLE IF NOT EXISTS incidents (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      workspace_id TEXT NOT NULL,
      project_id TEXT NOT NULL,
      timestamp TEXT DEFAULT CURRENT_TIMESTAMP,
      type TEXT NOT NULL,
      agent TEXT NOT NULL,
      provider TEXT NOT NULL,
      status_code INTEGER,
      message TEXT NOT NULL,
      goal_id TEXT
    )
  `);
} catch (err) {
  console.warn(`[SCAR] SQLite database unavailable (${err.message}). Falling back to JSON incident logs.`);
  useFallback = true;
  fs.mkdirSync(path.dirname(fallbackFile), { recursive: true });
  if (fs.existsSync(fallbackFile)) {
    try {
      fallbackIncidents = JSON.parse(fs.readFileSync(fallbackFile, 'utf8'));
    } catch (_) {}
  }
}

export class SCAR {
  constructor(workspaceId = 'default', projectId = 'core') {
    this.workspaceId = workspaceId;
    this.projectId = projectId;

    this.knownFixes = [
      {
        regex: /fn\.apply is not a function|route file does not export a function|liveness check failed/i,
        label: 'ROUTE FACTORY EXPORT',
        hint: 'Route files MUST export a factory function: `module.exports = function(app) { app.get(...) }`. NEVER export a raw Express Router.'
      },
      {
        regex: /guardian rejection|banned call|forbidden execution/i,
        label: 'BANNED CALL EXHAUSTION',
        hint: 'eval(), exec(), or child_process.exec() are forbidden by security policies. Use spawn() with explicit args instead.'
      },
      {
        regex: /path traversal|path sanitization/i,
        label: 'PATH TRAVERSAL ATTEMPT',
        hint: 'No `..` or absolute paths. Always relative-to-projectRoot. Validate with path.resolve + startsWith check.'
      },
      {
        regex: /search block did not match|search.*replace.*invalid/i,
        label: 'INVALID SEARCH/REPLACE DIFF',
        hint: 'The SEARCH block must match the target file exactly. Double check whitespace and line endings, or use empty SEARCH blocks.'
      },
      {
        regex: /syntax check failed|unexpected end of input|unexpected token/i,
        label: 'SYNTAX COMPLIANCE ERROR',
        hint: 'Output must be valid syntax. Verify braces, brackets, parentheses, and commas. Never truncate output mid-block.'
      },
      {
        regex: /esm import|esmodule.*\.js file|cannot use import/i,
        label: 'ESM IMPORT IN COMMONJS',
        hint: 'This file runs in a CommonJS project. Use `const x = require(...)` and `module.exports = x`. Do NOT use `import` / `export` syntax.'
      },
      {
        regex: /rate limit|quota exceeded|429 status/i,
        label: 'API RATE LIMIT TRIGGER',
        hint: 'Provider is currently rate-limited. Wait or adjust model chains to route to cheaper/local providers first.'
      },
      {
        regex: /timeout|request exceeded \d+s/i,
        label: 'REQUEST TIMEOUT EXPIRED',
        hint: 'Inference took too long. Reduce output tokens size or request complexity. Check provider status.'
      }
    ];
  }

  /**
   * Clear all logged incidents (primarily for testing).
   */
  clearIncidents() {
    if (useFallback) {
      fallbackIncidents.length = 0;
      if (fs.existsSync(fallbackFile)) {
        try {
          fs.unlinkSync(fallbackFile);
        } catch (_) {}
      }
    } else if (db) {
      try {
        db.exec('DELETE FROM incidents');
      } catch (_) {}
    }
  }

  /**
   * Record an operational failure in the database.
   * @param {string} type 
   * @param {string} agent 
   * @param {string} provider 
   * @param {number|null} statusCode 
   * @param {string} message 
   * @param {string|null} [goalId] 
   * @returns {number}
   */
  recordIncident(type, agent, provider, statusCode, message, goalId = null) {
    if (useFallback) {
      const id = fallbackIncidents.length + 1;
      const incident = {
        id,
        workspace_id: this.workspaceId,
        project_id: this.projectId,
        timestamp: new Date().toISOString(),
        type,
        agent,
        provider,
        status_code: statusCode,
        message,
        goal_id: goalId
      };
      fallbackIncidents.push(incident);
      try {
        fs.writeFileSync(fallbackFile, JSON.stringify(fallbackIncidents, null, 2), 'utf8');
      } catch (e) {
        console.error(`[SCAR] Write fallback failed: ${e.message}`);
      }
      return id;
    }

    try {
      const stmt = db.prepare(`
        INSERT INTO incidents (workspace_id, project_id, type, agent, provider, status_code, message, goal_id)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
      `);
      const info = stmt.run(this.workspaceId, this.projectId, type, agent, provider, statusCode, message, goalId);
      return info.lastInsertRowid;
    } catch (e) {
      console.error(`[SCAR] Record incident failed: ${e.message}`);
      return -1;
    }
  }

  /**
   * Fetch recent logged failures.
   * @param {number} [limit] 
   * @returns {Array<Object>}
   */
  getRecentScars(limit = 10) {
    if (useFallback) {
      return fallbackIncidents
        .filter(inc => inc.workspace_id === this.workspaceId && inc.project_id === this.projectId)
        .slice(-limit)
        .reverse();
    }

    try {
      const stmt = db.prepare(`
        SELECT type, agent, provider, status_code as status_code, message, goal_id as goal_id, timestamp
        FROM incidents
        WHERE workspace_id = ? AND project_id = ?
        ORDER BY id DESC
        LIMIT ?
      `);
      return stmt.all(this.workspaceId, this.projectId, limit);
    } catch (e) {
      console.error(`[SCAR] Fetch scars failed: ${e.message}`);
      return [];
    }
  }

  /**
   * Scans recent failures and counts occurrences of matching signatures.
   * @param {Array<Object>} scars 
   * @returns {Array<Object>}
   */
  detectFailurePatterns(scars) {
    if (!scars || scars.length === 0) return [];
    const patternCounts = new Map();

    for (const scar of scars) {
      const message = scar.message || '';
      for (const fix of this.knownFixes) {
        if (fix.regex.test(message)) {
          const label = fix.label;
          if (!patternCounts.has(label)) {
            patternCounts.set(label, { count: 0, hint: fix.hint });
          }
          patternCounts.get(label).count += 1;
        }
      }
    }

    const repeated = [];
    for (const [label, data] of patternCounts.entries()) {
      if (data.count >= 2) {
        repeated.push({
          pattern: label,
          count: data.count,
          hint: data.hint
        });
      }
    }

    return repeated.sort((a, b) => b.count - a.count);
  }

  /**
   * Inject warning blocks if identical failures repeat.
   * @param {string} prompt 
   * @param {Array<Object>} scars 
   * @returns {string}
   */
  injectRepeatGuard(prompt, scars) {
    const repeated = this.detectFailurePatterns(scars);
    if (repeated.length === 0) return prompt;

    let warningBlock = '\n\n### !!! REPEAT FAILURE GUARD — DO NOT REPEAT THESE ERRORS !!!\n';
    warningBlock += 'The following errors have occurred MULTIPLE times in this session. Prioritize fixing them:\n\n';

    for (const p of repeated) {
      warningBlock += `▶ Pattern: ${p.pattern} (Failed ${p.count} times)\n`;
      warningBlock += `  Fix Instruction: ${p.hint}\n\n`;
    }

    warningBlock += '### !!! END REPEAT FAILURE GUARD !!!\n\n';
    return warningBlock + prompt;
  }
}

File: python/engines/memory/scar.py
"""
Sovereign Critical Action Record (SCAR) database and Repeat Failure Guard.
Prevents AI agents from repeating identical errors in loops.
"""
import sqlite3
import os
import re
import logging
from datetime import datetime
from typing import List, Dict, Any, Optional

logger = logging.getLogger(__name__)

class SCAR:
    """Manages SQLite incident logging and Repeat Failure Guard prompt injections."""
    
    def __init__(self, db_path: str = "data/scars.db", workspace_id: str = "default", project_id: str = "core"):
        self.db_path = db_path
        self.workspace_id = workspace_id
        self.project_id = project_id
        os.makedirs(os.path.dirname(self.db_path), exist_ok=True)
        self._init_db()
        
        # Known failure signatures and their corresponding fix instructions
        self.known_fixes = [
            {
                "regex": r"fn\.apply is not a function|route file does not export a function|liveness check failed",
                "label": "ROUTE FACTORY EXPORT",
                "hint": "Route files MUST export a factory function: `module.exports = function(app) { app.get(...) }`. NEVER export a raw Express Router."
            },
            {
                "regex": r"guardian rejection|banned call|forbidden execution",
                "label": "BANNED CALL EXHAUSTION",
                "hint": "eval(), exec(), or child_process.exec() are forbidden by security policies. Use spawn() with explicit args instead."
            },
            {
                "regex": r"path traversal|path sanitization",
                "label": "PATH TRAVERSAL ATTEMPT",
                "hint": "No `..` or absolute paths. Always relative-to-projectRoot. Validate with path.resolve + startsWith check."
            },
            {
                "regex": r"search block did not match|search.*replace.*invalid",
                "label": "INVALID SEARCH/REPLACE DIFF",
                "hint": "The SEARCH block must match the target file exactly. Double check whitespace and line endings, or use empty SEARCH blocks."
            },
            {
                "regex": r"syntax check failed|unexpected end of input|unexpected token",
                "label": "SYNTAX COMPLIANCE ERROR",
                "hint": "Output must be valid syntax. Verify braces, brackets, parentheses, and commas. Never truncate output mid-block."
            },
            {
                "regex": r"esm import|esmodule.*\.js file|cannot use import",
                "label": "ESM IMPORT IN COMMONJS",
                "hint": "This file runs in a CommonJS project. Use `const x = require(...)` and `module.exports = x`. Do NOT use `import` / `export` syntax."
            },
            {
                "regex": r"rate limit|quota exceeded|429 status",
                "label": "API RATE LIMIT TRIGGER",
                "hint": "Provider is currently rate-limited. Wait or adjust model chains to route to cheaper/local providers first."
            },
            {
                "regex": r"timeout|request exceeded \d+s",
                "label": "REQUEST TIMEOUT EXPIRED",
                "hint": "Inference took too long. Reduce output tokens size or request complexity. Check provider status."
            }
        ]

    def _init_db(self):
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS incidents (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    workspace_id TEXT NOT NULL,
                    project_id TEXT NOT NULL,
                    timestamp TEXT DEFAULT CURRENT_TIMESTAMP,
                    type TEXT NOT NULL,
                    agent TEXT NOT NULL,
                    provider TEXT NOT NULL,
                    status_code INTEGER,
                    message TEXT NOT NULL,
                    goal_id TEXT
                )
            """)
            conn.commit()

    def record_incident(
        self,
        incident_type: str,
        agent: str,
        provider: str,
        status_code: Optional[int],
        message: str,
        goal_id: Optional[str] = None
    ) -> int:
        """Record an operational failure in the database."""
        try:
            with sqlite3.connect(self.db_path) as conn:
                cursor = conn.cursor()
                cursor.execute(
                    """
                    INSERT INTO incidents (workspace_id, project_id, type, agent, provider, status_code, message, goal_id)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    (self.workspace_id, self.project_id, incident_type, agent, provider, status_code, message, goal_id)
                )
                conn.commit()
                return cursor.lastrowid
        except Exception as e:
            logger.error(f"Failed to record incident to SCAR: {e}")
            return -1

    def get_recent_scars(self, limit: int = 10) -> List[Dict[str, Any]]:
        """Fetch recent logged failures for the current workspace & project."""
        try:
            with sqlite3.connect(self.db_path) as conn:
                conn.row_factory = sqlite3.Row
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT type, agent, provider, status_code, message, goal_id, timestamp
                    FROM incidents
                    WHERE workspace_id = ? AND project_id = ?
                    ORDER BY id DESC
                    LIMIT ?
                    """,
                    (self.workspace_id, self.project_id, limit)
                )
                return [dict(row) for row in cursor.fetchall()]
        except Exception as e:
            logger.error(f"Failed to fetch scars: {e}")
            return []

    def detect_failure_patterns(self, scars: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Scans recent failures and counts occurrences of matching signatures."""
        if not scars:
            return []
            
        pattern_counts = {}
        for scar in scars:
            message = scar.get("message", "")
            for fix in self.known_fixes:
                if re.search(fix["regex"], message, re.IGNORECASE):
                    label = fix["label"]
                    if label not in pattern_counts:
                        pattern_counts[label] = {
                            "count": 0,
                            "hint": fix["hint"]
                        }
                    pattern_counts[label]["count"] += 1
                    
        # Filter patterns that occur >= 2 times
        repeated = []
        for label, data in pattern_counts.items():
            if data["count"] >= 2:
                repeated.append({
                    "pattern": label,
                    "count": data["count"],
                    "hint": data["hint"]
                })
                
        # Sort by most frequent first
        repeated.sort(key=lambda x: x["count"], reverse=True)
        return repeated

    def inject_repeat_guard(self, prompt: str, scars: List[Dict[str, Any]]) -> str:
        """Injects a high-priority warning block into prompts if identical failures repeat."""
        repeated_patterns = self.detect_failure_patterns(scars)
        if not repeated_patterns:
            return prompt
            
        warning_block = "\n\n### !!! REPEAT FAILURE GUARD — DO NOT REPEAT THESE ERRORS !!!\n"
        warning_block += "The following errors have occurred MULTIPLE times in this session. Prioritize fixing them:\n\n"
        
        for p in repeated_patterns:
            warning_block += f"▶ Pattern: {p['pattern']} (Failed {p['count']} times)\n"
            warning_block += f"  Fix Instruction: {p['hint']}\n\n"
            
        warning_block += "### !!! END REPEAT FAILURE GUARD !!!\n\n"
        
        # Inject the warning block right at the top of the prompt payload
        return warning_block + prompt
